docs/developer/testing-in-fedbiomed.md

```diff
@@ -0,0 +1,167 @@
+- Test the methods that are overwritten by the class being tested.
+- When testing extended methods, please follow the principles in [Item 3 - Focus on smaller units](#3-focus-on-smaller-units).
+
+### 7. Testing Wrappers
+
+#### 7.1 Wrappers that wrap objects with a COMMON interface
+- Focus on wrapper methods rather than the wrapped objects. As objects that can be wrapped by the wrapper have a common interface, use a single object instead of applying the same test for each of them.
+- If the wrapped object is complex and requires different states, consider creating a fake object, and follow the section [Using Fake objects/classes](#8-using-fake-objects-classes) to determine if a fake object is necessary.
+
+#### 7.2 Wrappers that wrap objects with DIFFERENT interfaces
+- If the wrapper wraps different objects with different interfaces (methods), you can write tests for each wrapped object.
+- Follow the same logic as in "7.1" for mocking or using fake objects.
+
+### 8. Using Fake or Stub objects/classes
+- Fake or Stub objects require coherency between the real implementation and the fake one. Therefore, avoid using fake objects as much as possible because they can be difficult to maintain.
+- Please answer the following questions before deciding to create a fake class:
+    - 1. Is the object passed to a function or class as an argument?
+    - 2. Is the object dependent on the system or external actions (e.g., database)?
+    - 3. Does the object have too many dependencies on other modules of the library?
+- If 1. and 2. or 3. are "yes," feel free to create a fake object.
+- If 1. is "no" and the others are "yes," consider using mocking instead of fake or stub objects.
+
+### 9. Using Mocked objects
+- Mocking helps isolate the test from the environment and easily test different scenarios of a single function.
+- Pay attention to using the `spec`, `spec_set`, or `auto_spec` functionalities of `unittest.mock.Mock`. These functionalities help mock objects that respect the specifications of the real class, so the mock will be updated if an API of the real class changes.
+- Using too many mock objects may not be a good idea because:
+    - It may increase the complexity of the test function.
+    - It may isolate the test too much from the remaining parts of the software, resulting in invalid tests.
+
+- Before deciding to mock an object, please answer the following questions:
+    - 1. Is the object instantiated or used directly within the module or class being tested?
+    - 2. Is the object dependent on the system or external actions (e.g., database, network etc)?
+    - 3. Does the object have too many dependencies on other modules of the library?
+    - 4. Does the method being tested require many results or states of the object that needs to be mocked?
+    - 5. Is mocking saving you a lot of time and reducing complexity?
+    - 6. 
```

Comment:
there is a spurious item here I think

Reply:
I removed the item but the item number left there :)